src/common/util/queueImmut.ml

```diff
@@ -0,0 +1,126 @@
+(* A Queue (FIFO, imutable) implementation, using two Lists for efficient push and pop handling.
+
+   Example Queue: remove here, 1, 2, 3, 4, add here
+   Representation as Queue: Queue([1,2], [4,3]) or Queue([1], [4,3,2]) or Queue([1,2,3,4], []) or ...
+*)
+
+exception Empty
+
+type 'a t = Queue of 'a list * 'a list
+
+(** Return a new queue, initially empty. *)
+let create () = Queue ([], [])
+let length (Queue (q_first, q_last)) = List.length q_first + List.length q_last
+
+(** Return [true] if the given queue is empty, [false] otherwise. *)
+let is_empty = function
+  | Queue ([], []) -> true
+  | _ -> false
+
+(** Discard all elements from a queue. *)
+let clear (Queue (q_first, q_last)) = Queue ([], [])
+
+(** The queue is internally stored as two lists. [get_first_list q] with [q] equal to [Queue (q_first, q_last)] returns [q_first]*)
+let get_first_list (Queue (q_first, q_last)) = q_first
+
+(** The queue is internally stored as two lists. [get_first_list q] with [q] equal to [Queue (q_first, q_last)] returns [q_last]*)
+let get_last_list (Queue (q_first, q_last)) = q_last
+
+(** Converts a queue into a list. *)
+let list_of_queue = function 
+  | Queue (q_first, []) -> q_first
+  | Queue (q_first, q_last) -> q_first @ List.rev q_last
+
+(** Converts a list into a queue. *)
+let queue_of_list list = Queue (list,[])
+
+let rec equal eq q1 q2 =
+  match q1, q2 with
+  | Queue ([], []), Queue ([], []) -> true
+  | Queue ([], []), Queue ([], x) | Queue ([], []), Queue (x, []) 
+  | Queue (x, []), Queue ([], []) | Queue ([], x), Queue ([], []) -> false
+  | Queue ([], q1_last), Queue (q2_first, q2_last) -> equal eq (Queue (List.rev q1_last, [])) (Queue (q2_first, q2_last))
+  | Queue (q1_first, q1_last), Queue ([], q2_last) -> equal eq (Queue (q1_first, q1_last)) (Queue (List.rev q2_last, []))
+  | Queue (x1::q1_first, q1_last), Queue (x2::q2_first, q2_last) -> eq x1 x2 && equal eq (Queue (q1_first, q1_last)) (Queue (q2_first, q2_last)) 
+
+let rec compare cmp q1 q2 = 
+  match q1, q2 with
+  | Queue ([], []), Queue ([], []) -> 0
+  | Queue ([], []), Queue ([], x) | Queue ([], []), Queue (x, []) -> -1
+  | Queue (x, []), Queue ([], []) | Queue ([], x), Queue ([], []) -> 1 
+  | Queue ([], q1_last), Queue (q2_first, q2_last) -> compare cmp (Queue (List.rev q1_last, [])) (Queue (q2_first, q2_last))
+  | Queue (q1_first, q1_last), Queue ([], q2_last) -> compare cmp (Queue (q1_first, q1_last)) (Queue (List.rev q2_last, []))
+  | Queue (x1::q1_first, q1_last), Queue (x2::q2_first, q2_last) -> 
+    let c = cmp x1 x2 in
+    if c <> 0 then c 
+    else compare cmp (Queue (q1_first, q1_last)) (Queue (q2_first, q2_last)) 
```

Comment:
Is there some possibility for code reuse between `equal` and `compare`? It seems like most of the code is the same?